Improvement/1558 encrypt k8s secrets on disk

jbertran opened the pull request:
**Component**: salt/apiserver

**Context**:

Currently the K8s `Secret` objects are saved as plain text in etcd, which is perfectible™.

**Summary**:

* Set up an encryption configuration state pulled by the `kube-apiserver` installation state, which is itself pulled by the role setup.
* Use the private pillar to ensure that the encryption key is generated once and only once, and then replicated across the other nodes running `kube-apiserver`.
* Make `kube-apiserver` use that encryption configuration by mounting it into the container and adding the relevant flag to its launch options.

**Acceptance criteria**: 

1. Get a cluster up with this branch, go into any machine of the cluster and switch to root
2. Create a secret: `kubectl create secret generic secret1 -n default --from-literal=mykey=mydata`
3. Verify the secret is legible: `kubectl get secret secret1 -o yaml` - the data is encoded in base64.
4. Verify the secret is encrypted in etcd:
     a. exec into the etcd container: `kubectl -ti etcd-bootstrap sh -n kube-system --kubeconfig /etc/kubernetes/admin.conf`
     b. dump the secret: `ETCDCTL_API=3 etcdctl get /registry/secrets/default/secret1 --cert=/etc/kubernetes/pki/etcd/server.crt --key=/etc/kubernetes/pki/etcd/server.key --cacert=/etc/kubernetes/pki/etcd/ca.crt | hexdump -C`
     c. The data should be illegible, and should be prefixed by `:enc:aescbc:v1`

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1558 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello jbertran,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1558-encrypt-k8s-secrets-on-disk` into
`development/2.0`
* [w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.1` 
* [w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.2` 
* [w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.3` 
* [w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

jbertran committed:
salt: provide encryption key for secrets in `private` ext pillar

This pillar section is populated from the contents of
/etc/metalk8s/crypt/encryption.key

jbertran committed:
salt: provide a cryptconfig state for apiserver

This state is responsible for generating the configuration file that
`kube-apiserver` takes as argument of its launch command.

If the `encryption` key is unavailable from the pillar, the configuration
file is generated by the state. This is guaranteed to be on the bootstrap
node due to the fact that it is always the first node to be brought to
highstate.

jbertran committed:
salt: use the `cryptconfig` formula to activate on-disk encryption

bert-e commented:
# History mismatch



Merge commit #738641803edce566531e07c7a05e419a97171278 on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jbertran commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1558-encrypt-k8s-secrets-on-disk` into
`development/2.0`
* [w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.1` 
* [w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.2` 
* [w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.3` 
* [w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #243c575169f66e094d328f7f1826656241b64246 on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #b829d44f8d08806a117bf8697481f2de8e086cde on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jbertran commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1558-encrypt-k8s-secrets-on-disk` into
`development/2.0`
* [w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.1` 
* [w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.2` 
* [w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.3` 
* [w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

jbertran commented:
/create_pull_requests

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1558-encrypt-k8s-secrets-on-disk` into
`development/2.0`
* [w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.1` (pull request #1647)
* [w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.2` (pull request #1648)
* [w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.3` (pull request #1649)
* [w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.4` (pull request #1650)



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **create_pull_requests**

charlesprost commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -1,8 +1,10 @@
 
 
 SA_PRIVATE_KEY_PATH = "/etc/kubernetes/pki/sa.key"
+APISERVER_KEY_PATH = "/etc/metalk8s/apiserver.key"
```
```suggestion
APISERVER_KEY_PATH = "/etc/metalk8s/crypt/apiserver.key"
```
So we have a clear distinction between PKI stuff and encryption stuff

jbertran replied:
I was unsure about the path, good suggestion, done.

charlesprost commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -28,6 +30,23 @@ def _read_sa_private_key():
 
+def _read_apiserver_key():
+    # FIXME: we only have access to this file because:
+    #   1- Salt master Pod mounts all of /etc/kubernetes
```
You are talking about `/etc/kubernetes` but the key file is stored under `/etc/metalk8s`.

jbertran replied:
Oof, indeed. The remark still holds, edited the path :)

slaperche-scality commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -28,6 +30,23 @@ def _read_sa_private_key():
+    #   2- Salt master Pod runs on the 'bootstrap' minion, which is, by chance,
+    #      also the initial master node
+
+    if not os.path.isfile(APISERVER_KEY_PATH):
```
TOCTOU:, use a try/except (you can have two except, one for not found and one for the others) below instead.

jbertran replied:
cargo culting is bad, mmmkay?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality







*The following options are set:* **create_pull_requests**

bert-e commented:
# History mismatch



Merge commit #34143643eb71fe683ef0006f4c440d2a322ba807 on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_pull_requests**

bert-e commented:
# History mismatch



Merge commit #df6417d3c8b552b6fab038ce76d36377326169e0 on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_pull_requests**

jbertran commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_pull_requests**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1558-encrypt-k8s-secrets-on-disk` into
`development/2.0`
* [w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.1` (pull request #1654)
* [w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.2` (pull request #1655)
* [w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.3` (pull request #1656)
* [w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.4` (pull request #1657)



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality







*The following options are set:* **create_pull_requests**

jbertran committed:
pillar: better error handling for file contents loaded by pillar

Checking that a given path is a file before reading is risky, managing
the exception that would occur when file at that path is read is not.

We use a helper function because DRY, but keep the function definitions
in order to keep the corresponding comments.

slaperche-scality commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -35,14 +36,14 @@ def _read_apiserver_key():
     #      also the initial master node
 
-    if not os.path.isfile(APISERVER_KEY_PATH):
+    try:
```
Actually, the real equivalent would be:

```python
    try:
        with open(APISERVER_KEY_PATH, "r") as config_file:
            encryption_key = config_file.read()
    except FileNotFoundError:
        return {"_errors": [
            "Missing secrets encryption key to share with master node."
        ]}
```

Just catch file missing with a proper errors, and let other exceptions bubble up.

slaperche-scality replied:
Meh, no `FileNotFoundError` in Python 2, so it would be more like (need `import errno`):

```python
    try:
        with open(APISERVER_KEY_PATH, "r") as config_file:
            encryption_key = config_file.read()
    except IOError as exn:
        if exn.errno == errno.ENOENT:
            return {"_errors": [
                "Missing secrets encryption key to share with master node."
            ]}
```

bert-e commented:
# History mismatch



Merge commit #cc1765e7a4cdaa7eef7ccdf43104d7f720dbd4fd on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_pull_requests**

slaperche-scality commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -14,36 +14,39 @@ def __virtual__():
+        with open(key_path, "r") as key_file:
+            key_data = key_file.read()
+    except IOError as exn:
+        # File not found
```
IMHO, `ENOENT` is clear enough (that's the point of importing `errno` to use the named constant instead of `2`)

slaperche-scality commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -14,36 +14,39 @@ def __virtual__():
+            ]}
+        # Other failures to read the file
+        err_msg = "Unable to read {0} key file contents at {1}."
+        return {"_errors": [err_msg.format(key_name, key_path)]}
```
could be good to have the exception message somewhere here.

slaperche-scality commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -14,36 +14,39 @@ def __virtual__():
     return __virtualname__
 
 
+def _read_private_key(key_name, key_path):
```
Could explain in the commit message why this reword (TOCTOU, obvious in the context of this review but not when finding the commit alone with a git blame)

jbertran replied:
Done

jbertran commented:
/reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**





*The following options are set:* **create_pull_requests**

bert-e commented:
# History mismatch



Merge commit #cc1765e7a4cdaa7eef7ccdf43104d7f720dbd4fd on the integration branch
`w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk` is merging a branch which is neither the current
branch `improvement/1558-encrypt-k8s-secrets-on-disk` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1558-encrypt-k8s-secrets-on-disk` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_pull_requests**

jbertran commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_pull_requests**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1558-encrypt-k8s-secrets-on-disk` into
`development/2.0`
* [w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.1` (pull request #1659)
* [w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.2` (pull request #1660)
* [w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.3` (pull request #1661)
* [w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk) will be merged into `development/2.4` (pull request #1662)



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **create_pull_requests**

jbertran commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/ab97d99a22888ef651b810d82012f8df295f5b40) did not succeed in branch w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk.




*The following options are set:* **create_pull_requests, approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/809df7d904aaa3cb2d809738b906922de9646c88) did not succeed in branch w/2.2/improvement/1558-encrypt-k8s-secrets-on-disk.




*The following options are set:* **create_pull_requests, approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/ab97d99a22888ef651b810d82012f8df295f5b40) did not succeed in branch w/2.4/improvement/1558-encrypt-k8s-secrets-on-disk.




*The following options are set:* **create_pull_requests, approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **create_pull_requests, approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye jbertran.
